feat(script): add commands & entrypoint

byeongsu-hong opened the pull request:
* context
    * make-agent-config: context -> hyperlane agent config
* contract
    * list: show all contracts
    * test-dispatch: dispatch tester (amount is fixed at this time)
* deploy: deploys contracts by configuration (default = `./config.yaml`)
* migrate: migrates contracts if contract have code changes
    * backlogs
         1. fetch contract codes from remote
         3. ledger support
         4. export to `x/gov` proposal messages
         5. export to DAODAO proposal messages
* upload
    * local: uploads local contract codes. default artifact path is `./artifacts`
    * remote: fetches contract codes from Github Release by tag like `v0.0.6-rc8`, and uploads it
    * remote-list: show list of available release
* wallet
    * create account and exports mnemonic, privateKey
    * derive address of network (bech32 prefix) from injected `mnemonic` or `privateKey`

byeongsu-hong committed:
feat: add shared utilities

byeongsu-hong committed:
fix: correct representation of `instant finality`

byeongsu-hong committed:
fix: also actual value

byeongsu-hong committed:
feat: add deploy script for hook

byeongsu-hong committed:
feat: add ism deployer

byeongsu-hong committed:
feat: add igp deployer

byeongsu-hong committed:
feat: export everything

byeongsu-hong committed:
feat: add commands

byeongsu-hong committed:
feat: main entrypoint

nambrot commented on script/commands/contract.ts:
```diff
@@ -0,0 +1,48 @@
+});
+
+contractCmd.command("list").action(() => {
+  console.log("Available contracts:".green);
```
Does this syntax work, never seen this before

byeongsu-hong replied:
Yes, it works

nambrot commented on script/commands/contract.ts:
```diff
@@ -0,0 +1,48 @@
+        dispatch: {
+          dest_domain: parseInt(destDomain),
+          recipient_addr: addPad(
+            recipientAddr.startsWith("0x")
```
Nit: Move this into a utility function

nambrot commented on script/commands/deploy.ts:
```diff
@@ -0,0 +1,190 @@
+  ]);
+
+  saveContext(opts.networkId, ctx);
+  saveAgentConfig(getNetwork(opts.networkId), ctx);
```
Don't you have to await here? Should probably have a lint rule to not allow for this situation

byeongsu-hong replied:
Oh my, will fix it right away

nambrot commented on script/commands/deploy.ts:
```diff
@@ -0,0 +1,190 @@
+): Promise<ContextDeployments["core"]> => {
+  const { hrp, domain } = getNetwork(networkId);
+
+  const log = (v: string) => console.log("[core]".green, v);
```
This seems like another utility function?

byeongsu-hong replied:
https://github.com/many-things/cw-hyperlane/pull/109/files#diff-e2146dde0d27921365622c6faf7bff74f6eda8dfc6bed377288782c1ca5c3472 refactored this here

nambrot commented on script/commands/deploy.ts:
```diff
@@ -0,0 +1,190 @@
+      domain,
+      owner: client.signer,
+    }));
+  if (preload?.mailbox) log(`${deployment.mailbox.type} already deployed`);
```
Shouldn't this log line be above?

byeongsu-hong replied:
You're right

nambrot commented on script/commands/deploy.ts:
```diff
@@ -0,0 +1,190 @@
+    (await deployIsm(
+      ctx,
+      client,
+      config.deploy.ism instanceof Array
```
Very deep indentation, move this out IMO

nambrot commented on script/commands/deploy.ts:
```diff
@@ -0,0 +1,190 @@
+            validators: config.deploy.ism
+              .map((domain) => ({
+                domain,
+                addrs: [client.signer_addr],
```
Is this intended?

byeongsu-hong replied:
Yes, but we decided to remove default seting of ism, it is redundnat

nambrot commented on script/commands/deploy.ts:
```diff
@@ -0,0 +1,190 @@
+        ? {
+            // default ism (multisig)
+            type: "multisig",
+            owner: client.signer,
```
Ditto?

byeongsu-hong replied:
Same as above

nambrot commented on script/commands/deploy.ts:
```diff
@@ -0,0 +1,190 @@
+      networkId,
+      ctx,
+      client,
+      config.deploy.hooks.default || { type: "mock" }
```
Let's not default to mock here, but throw an error, too hard to debug if you forgot to set the type

byeongsu-hong replied:
Agreed, let's get rid of this.

nambrot commented on script/commands/deploy.ts:
```diff
@@ -0,0 +1,190 @@
+      networkId,
+      ctx,
+      client,
+      config.deploy.hooks.required || { type: "mock" }
```
Ditto

nambrot commented on script/commands/deploy.ts:
```diff
@@ -0,0 +1,190 @@
+import { deployContract, executeMultiMsg } from "../shared/contract";
+import { saveAgentConfig } from "../shared/agent";
+
+export const deployCmd = new Command("deploy")
```
Can be done in a separate command, but IMO it would be valuable to be able to deploy ISMs and hooks by itself

byeongsu-hong replied:
Yes, i agree about it. Let's leave this unified for now, and i'll separate it later.

nambrot commented on script/commands/migrate.ts:
```diff
@@ -0,0 +1,149 @@
+
+  const { ctx, client } = CONTAINER.get(Dependencies);
+
+  const flatten = [
```
Better variable name please like `flattenedContracts`

nambrot commented on script/commands/migrate.ts:
```diff
@@ -0,0 +1,149 @@
+export const migrateCmd = new Command("migrate")
+  .description("Migrate contracts")
+  .configureHelp({ showGlobalOptions: true })
+  .option("-c --contracts <contracts...>", "specify contracts to migrate")
```
How do you pass multiple? comma separated?

nambrot replied:
Be clear that you are looking for types here

byeongsu-hong replied:
https://github.com/tj/commander.js?tab=readme-ov-file#variadic-option like this

nambrot commented on script/commands/migrate.ts:
```diff
@@ -0,0 +1,149 @@
+    )
+  ) as Record<ContractNames, CodeDetails>;
+
+  const toMigrate = withContractInfo.filter(
```
what does this do?

byeongsu-hong replied:
comparing a checksum of code that willing to migrate and actual one

nambrot commented on script/commands/migrate.ts:
```diff
@@ -0,0 +1,149 @@
+    return;
+  }
+
+  if (!(await askQuestion("Do you want to proceed? (y/n)"))) {
```
Should we log out what's gonna happen to allow the user to know?

byeongsu-hong replied:
Agree!

nambrot commented on script/commands/migrate.ts:
```diff
@@ -0,0 +1,149 @@
+  console.log("Proceeding to migrate...");
+
+  for (const migrate of toMigrate) {
+    const res = await client.wasm.migrate(
```
Log what's happening

nambrot commented on script/commands/migrate.ts:
```diff
@@ -0,0 +1,149 @@
+
+  for (const migrate of toMigrate) {
+    const res = await client.wasm.migrate(
+      client.signer,
```
When the owner is a DAO DAO contract, can we just log out the calls?

byeongsu-hong replied:
Yes, we can export these msgs like this format.

```
{
    "wasm": {
        "migrate": {
            "contract_addr": "~~",
            "new_code_id": 1234,
            "msg": "<base64-encoded-msg>"
        }
    }
}
```

nambrot commented on script/commands/upload.ts:
```diff
@@ -0,0 +1,216 @@
+  .command("local")
+  .description("upload artifacts from local")
+  .option("-a --artifacts <artifacts dir>", "artifacts", defaultArtifactPath)
+  .action(async (_, cmd) => upload(cmd.optsWithGlobals()));
```
Nit: async is superflous

byeongsu-hong committed:
Squashed commit of the following:

commit fad8f9e59a9de9b25028fab89c6c851d50d0dd7d
Merge: 1639e2f eb77beb
Author: byeongsu-hong <hong@byeongsu.dev>
Date:   Fri Mar 1 23:21:15 2024 +0900

    Merge branch 'eddy/pick-docs' into eddy/pick-cw-hpl

commit 1639e2fd0983c9bdc3b9814b7c8b314bd803a7d7
Author: byeongsu-hong <hong@byeongsu.dev>
Date:   Fri Mar 1 23:17:08 2024 +0900

    remove ts/sdk

commit d73647b4b073e1f142b13fd44da723cf7dcba153
Author: byeongsu-hong <hong@byeongsu.dev>
Date:   Fri Mar 1 23:15:19 2024 +0900

    Revert "Merge branch 'main' into eddy/pick-cw-hpl"

    This reverts commit 82fc10fc57ccf868ae172944c34974fa8917a48c, reversing
    changes made to 6566e7ad0ee50198c46a463369d7dfececb9865f.

commit 82fc10fc57ccf868ae172944c34974fa8917a48c
Merge: 6566e7a be64967
Author: byeongsu-hong <hong@byeongsu.dev>
Date:   Fri Mar 1 23:14:05 2024 +0900

    Merge branch 'main' into eddy/pick-cw-hpl

commit be64967876be5e8b0de21d07bd3544ef48afdf41
Author: ByeongSu Hong <hong@byeongsu.dev>
Date:   Fri Mar 1 23:08:13 2024 +0900

    doc: documentation (#95)

    * wip

    * project structure & overview

    * toc

    * swap

    * fix(schema): reflect missing contracts

    * refactor(ts): migrate sdk - 1

    * refactor(script): pull back project setup to root

    * reorg script / cli

    * fix(script): make igp deployable

    * build: redeploy

    * build: stride hyperlane deployment

    * fix(hook): default gas denom

    * feat(script): add migrate command

    * chore: detailed ignore policy

    * feat(script): supprot mnemonic

    * chore(script): cleanup

    * chore: trailing comma

    * feat: add grpc endpoint as config

    * feat: generate agent config

    * feat(script): add test-dispatch

    * chore: redundant args

    * docs: owner -> <signer>

    * build: yarn berry

    * env: yarn

    * env: add cw-hpl command

    * feat: mailbox null-check

    * fix: prune imports

    * docs: wip guide & example

    * fix: rename compose file

    * fix: use localosmosis key

    * refactor: restruct guide

    * fix: remove DATA_PATH

    * fix: add `yarn install`

    * fix: handle rest endpoint not working

    * fix: use localwasmd

    * fix: hrp

    * feat: add wallet command

    * feat: wallet generator

    * fix: use osmosis testnet

    * line spacing

    * fix: xor

    * fix: split length

    * feat: more info to replace

    * fix: osmosis testnet network name

    * chore: ignore example

    * feat: add test recipient

    * fix: trouble shooting

    * docs: done

    * docs: update README.md

    * fix: line break

    * chore: remove testnet contexts

    * wip: neutron deployment

    * feat: context -> agent config

    * feat: apply review changes (#97)

    * docs: add context / example

    * fix: real instant finality

    * fix: typo

commit e8716c39d1ef82f8b641736acb73b62c74bd2990
Author: ByeongSu Hong <hong@byeongsu.dev>
Date:   Fri Mar 1 19:57:30 2024 +0900

    review: docs (#98)

    * wip

    * project structure & overview

    * toc

    * swap

    * docs: owner -> <signer>

    * build: yarn berry

    * env: yarn

    * env: add cw-hpl command

    * docs: wip guide & example

    * fix: line break

    * fix: rename compose file

    * fix: use localosmosis key

    * refactor: restruct guide

    * fix: remove DATA_PATH

    * fix: add `yarn install`

    * fix: use localwasmd

    * fix: hrp

    * fix: use osmosis testnet

    * line spacing

    * feat: more info to replace

    * fix: osmosis testnet network name

    * chore: ignore example

    * docs: done

    * docs: update README.md

    * merge: remove conflicts

    * docs: move to root

    * docs: remove outdated docs

commit eb77bebd6b33d514be379f40af8cffedf0e1369b
Author: byeongsu-hong <hong@byeongsu.dev>
Date:   Thu Feb 29 17:55:52 2024 +0900

    docs: remove outdated docs

commit 6a600c52dde686493a1c64e6c7b6d34ba4e382c6
Author: byeongsu-hong <hong@byeongsu.dev>
Date:   Thu Feb 29 17:54:38 2024 +0900

    docs: move to root

commit 68074f013565e92d4e596f93cb2eb52eac5551db
Author: ByeongSu Hong <hong@byeongsu.dev>
Date:   Thu Feb 29 02:20:50 2024 +0900

    refactor: ts schema (#99)

    * fix(schema): reflect missing contracts

    * refactor(ts): migrate sdk - 1

commit 5896800b212821e16411e25d0c618146f62b7d4c
Author: byeongsu-hong <hong@byeongsu.dev>
Date:   Wed Feb 28 17:16:32 2024 +0900

    merge: remove conflicts

commit adb18525cf60addcb6760d6950ed4e138d410408
Author: ByeongSu Hong <hong@byeongsu.dev>
Date:   Tue Feb 27 02:59:10 2024 +0900

    docs: update README.md

commit 19c6c60101b89e5be194a12bd3f007fddbd3a6e0
Author: byeongsu-hong <hong@byeongsu.dev>
Date:   Tue Feb 27 02:57:59 2024 +0900

    docs: done

commit 5cdecaa6ae19ee1e80579774180e4b2fdd1a9f94
Author: byeongsu-hong <hong@byeongsu.dev>
Date:   Tue Feb 27 02:32:56 2024 +0900

    chore: ignore example

commit 28a5e6917cb63f7b896fc254cf35a3651e5d63d6
Author: byeongsu-hong <hong@byeongsu.dev>
Date:   Tue Feb 27 02:21:21 2024 +0900

    fix: osmosis testnet network name

commit 57e78ea2e0c6bffd27f7d971304056a7db98f232
Author: byeongsu-hong <hong@byeongsu.dev>
Date:   Tue Feb 27 02:18:23 2024 +0900

    feat: more info to replace

commit 56e2700914607485f23cf7105dd6b6aa5ba73369
Author: byeongsu-hong <hong@byeongsu.dev>
Date:   Tue Feb 27 01:35:03 2024 +0900

    line spacing

commit b707b6748bdfa7ac5c7443e90d6e84f40504b987
Author: byeongsu-hong <hong@byeongsu.dev>
Date:   Tue Feb 27 01:32:35 2024 +0900

    fix: use osmosis testnet

commit aa53d70dbf7a82cf88105aa2721b7e66713f8bf4
Author: byeongsu-hong <hong@byeongsu.dev>
Date:   Tue Feb 27 00:38:22 2024 +0900

    fix: hrp

commit bc1b84f78f4078c3cf515d455d1ebe38459680aa
Author: byeongsu-hong <hong@byeongsu.dev>
Date:   Tue Feb 27 00:36:03 2024 +0900

    fix: use localwasmd

commit 7aaa751f71c6042c25d1ecc536e85995ac39a0d0
Author: byeongsu-hong <hong@byeongsu.dev>
Date:   Tue Feb 27 00:04:08 2024 +0900

    fix: add `yarn install`

commit ca25459e023c95eb20a2f76d83ce0d510bd88b0c
Author: byeongsu-hong <hong@byeongsu.dev>
Date:   Mon Feb 26 23:56:19 2024 +0900

    fix: remove DATA_PATH

commit 82b5f59f46054381c0331b5c074a49de8f07d254
Author: byeongsu-hong <hong@byeongsu.dev>
Date:   Mon Feb 26 23:49:25 2024 +0900

    refactor: restruct guide

commit df1a67cc1a68609254f299b0883149f27f623cff
Author: byeongsu-hong <hong@byeongsu.dev>
Date:   Mon Feb 26 22:32:48 2024 +0900

    fix: use localosmosis key

commit ef171651e991faabb5b1312c12e747049cf79189
Author: byeongsu-hong <hong@byeongsu.dev>
Date:   Mon Feb 26 22:26:51 2024 +0900

    fix: rename compose file

commit 1289770ac6960a9d6db41ad12426a93565cdd9e1
Author: byeongsu-hong <hong@byeongsu.dev>
Date:   Tue Feb 27 03:00:17 2024 +0900

    fix: line break

commit 2604486e096413bd846dd11e006c490c7f4e0b08
Author: byeongsu-hong <hong@byeongsu.dev>
Date:   Mon Feb 26 22:22:49 2024 +0900

    docs: wip guide & example

commit 2b4c5a1a5072841c2cc1b0eed30e0f6a9e8e9024
Author: byeongsu-hong <hong@byeongsu.dev>
Date:   Mon Feb 26 21:21:24 2024 +0900

    env: add cw-hpl command

commit e43cfb33cb19b1a5956111d3e5bbe662b019b8f0
Author: byeongsu-hong <hong@byeongsu.dev>
Date:   Mon Feb 26 21:21:14 2024 +0900

    env: yarn

commit 76313e89fdca069d274d094d9fa4a32a6b1344a4
Author: byeongsu-hong <hong@byeongsu.dev>
Date:   Mon Feb 26 20:49:06 2024 +0900

    build: yarn berry

commit 9dba0a4d93f447092326fa0f0db6f62b702d8ec5
Author: byeongsu-hong <hong@byeongsu.dev>
Date:   Mon Feb 26 20:44:20 2024 +0900

    docs: owner -> <signer>

commit 17200dc65d2f18c43dd98afba53c9dfb4de20782
Author: byeongsu-hong <hong@byeongsu.dev>
Date:   Wed Feb 14 18:50:16 2024 +0900

    swap

commit 50b01d70caf45028cde7c80a2f46e0e0884c69e1
Author: byeongsu-hong <hong@byeongsu.dev>
Date:   Wed Feb 14 18:42:09 2024 +0900

    toc

commit 6cde4c8da4bf937cf4ad15d0138cb9ce6c1f3c2e
Author: byeongsu-hong <hong@byeongsu.dev>
Date:   Wed Feb 14 18:29:22 2024 +0900

    project structure & overview

commit cab7199fb97355bdc2f333e56d0c7e47ba635cc1
Author: byeongsu-hong <hong@byeongsu.dev>
Date:   Wed Feb 14 18:22:41 2024 +0900

    wip

byeongsu-hong committed:
merge: cw-hpl (#110)

* refactor: ts schema (#99)

* fix(schema): reflect missing contracts

* refactor(ts): migrate sdk - 1

* review: docs (#98)

* wip

* project structure & overview

* toc

* swap

* docs: owner -> <signer>

* build: yarn berry

* env: yarn

* env: add cw-hpl command

* docs: wip guide & example

* fix: line break

* fix: rename compose file

* fix: use localosmosis key

* refactor: restruct guide

* fix: remove DATA_PATH

* fix: add `yarn install`

* fix: use localwasmd

* fix: hrp

* fix: use osmosis testnet

* line spacing

* feat: more info to replace

* fix: osmosis testnet network name

* chore: ignore example

* docs: done

* docs: update README.md

* merge: remove conflicts

* docs: move to root

* docs: remove outdated docs

* doc: documentation (#95)

* wip

* project structure & overview

* toc

* swap

* fix(schema): reflect missing contracts

* refactor(ts): migrate sdk - 1

* refactor(script): pull back project setup to root

* reorg script / cli

* fix(script): make igp deployable

* build: redeploy

* build: stride hyperlane deployment

* fix(hook): default gas denom

* feat(script): add migrate command

* chore: detailed ignore policy

* feat(script): supprot mnemonic

* chore(script): cleanup

* chore: trailing comma

* feat: add grpc endpoint as config

* feat: generate agent config

* feat(script): add test-dispatch

* chore: redundant args

* docs: owner -> <signer>

* build: yarn berry

* env: yarn

* env: add cw-hpl command

* feat: mailbox null-check

* fix: prune imports

* docs: wip guide & example

* fix: rename compose file

* fix: use localosmosis key

* refactor: restruct guide

* fix: remove DATA_PATH

* fix: add `yarn install`

* fix: handle rest endpoint not working

* fix: use localwasmd

* fix: hrp

* feat: add wallet command

* feat: wallet generator

* fix: use osmosis testnet

* line spacing

* fix: xor

* fix: split length

* feat: more info to replace

* fix: osmosis testnet network name

* chore: ignore example

* feat: add test recipient

* fix: trouble shooting

* docs: done

* docs: update README.md

* fix: line break

* chore: remove testnet contexts

* wip: neutron deployment

* feat: context -> agent config

* feat: apply review changes (#97)

* docs: add context / example

* fix: real instant finality

* fix: typo

* Revert "Merge branch 'main' into eddy/pick-cw-hpl"

This reverts commit 82fc10fc57ccf868ae172944c34974fa8917a48c, reversing
changes made to 6566e7ad0ee50198c46a463369d7dfececb9865f.

* remove ts/sdk

* merge: schema refactoring (#106)

* fix(schema): reflect missing contracts

* refactor(ts): migrate sdk - 1

* feat(script): initial config setup (#102)

feat: initial config setup

* merge: docs improvements (#108)

* wip

* project structure & overview

* toc

* swap

* docs: owner -> <signer>

* build: yarn berry

* env: yarn

* env: add cw-hpl command

* docs: wip guide & example

* fix: line break

* fix: rename compose file

* fix: use localosmosis key

* refactor: restruct guide

* fix: remove DATA_PATH

* fix: add `yarn install`

* fix: use localwasmd

* fix: hrp

* fix: use osmosis testnet

* line spacing

* feat: more info to replace

* fix: osmosis testnet network name

* chore: ignore example

* docs: done

* docs: update README.md

* merge: remove conflicts

* docs: move to root

* docs: remove outdated docs

byeongsu-hong committed:
fix: apply code review

byeongsu-hong committed:
chore: prevent logic duplication

byeongsu-hong committed:
fix: apply review changes
